src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorage.cs

```diff
@@ -24,94 +23,82 @@ public void Dispose()
 
         public void Init()
         {
-            foreach (var m in storageQueue.GetAllMessages())
-            {
-                var messageTypeString = m.Body as string;
-                var messageType = new MessageType(messageTypeString); //this will parse both 2.6 and 3.0 type strings
-                var subscriber = Deserialize(m.Label);
+            var messages = storageQueue.GetAllMessages()
+                .OrderByDescending(m => m.ArrivedTime)
+                .ThenBy(x => x.Id) // ensure same order of messages with same timestamp accross all endpoints
+                .ToArray();
 
-                entries.Add(new Entry
+            lock (lookup)
```

Comment:
hm just thinking this through: We call init from a feature startup task. At that point we already allow sending messages but we are not starting them concurrently yet (but I think there's a chance this might change?), so there is currently no risk of a concurrent subscriber lookup while initializing. Not sure whether the lock hurts though. In case we change this, it's hard to detect this race condition.

Also, now that we think about this, seems like publishing messages from a feature startup task may get into race conditions with `Init` anyway, should that really be a FST?